server/src/main/java/org/elasticsearch/action/bulk/Retry2.java

```diff
@@ -0,0 +1,286 @@
+     * @param maxNumberOfRetries This is the maximum number of times a BulkRequest will be retried
+     */
+    Retry2(int maxNumberOfRetries, ByteSizeValue maxBytesInFlight) {
+        this.logger = LogManager.getLogger(getClass());
+        this.maxNumberOfRetries = maxNumberOfRetries;
+        this.maxBytesInFlight = maxBytesInFlight.getBytes();
+    }
+
+    /**
+     * This method queues up the given BulkRequest. If there is no room on the queue, the listener is immediately notified of failure
+     * with an EsRejectedExecutionException. Otherwise, as soon as there is capacity consumer.accept(bulkRequest, actionListener) will be
+     * called (from another thread). If that call fails, the BulkRequest will be queued for retry based on this class's BackoffPolicy.
+     * @param consumer The consumer to which apply the request and listener. This consumer is expected to perform its work asynchronously
+     *                (that is, not block the thread from which it is called).
+     * @param bulkRequest The bulk request that should be executed.
+     * @param listener A listener that is invoked when the bulk request finishes or completes with an exception.
+     */
+    public void withBackoff(
+        BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer,
+        BulkRequest bulkRequest,
+        ActionListener<BulkResponse> listener
+    ) {
+        if (isClosing) {
+            listener.onFailure(new EsRejectedExecutionException("The bulk processor is closing"));
+            return;
+        }
+        List<BulkItemResponse> responsesAccumulator = new ArrayList<>();
+        logger.trace("Sending a bulk request with {} bytes in {} items", bulkRequest.estimatedSizeInBytes(), bulkRequest.requests.size());
+        /*
+         * Here we have a risk that multiple threads will do this check at once, and each will think it has room to run. So we could end
+         * up with slightly more bytes in flight than totalBytesInFlight. But this is acceptable risk, and preferable to using a lock or
+         * a while loop with compareAndSet.
+         */
+        long bytesInFlight = totalBytesInFlight.get();
+        if (bytesInFlight + bulkRequest.estimatedSizeInBytes() > maxBytesInFlight) {
+            listener.onFailure(
+                new EsRejectedExecutionException(
+                    "Cannot index request of size "
+                        + bulkRequest.estimatedSizeInBytes()
+                        + " because "
+                        + totalBytesInFlight.get()
+                        + " bytes are already in flight and the max is "
+                        + maxBytesInFlight
+                )
+            );
+        } else {
+            totalBytesInFlight.addAndGet(bulkRequest.estimatedSizeInBytes());
+            phaser.register();
+            consumer.accept(bulkRequest, new RetryHandler(bulkRequest, responsesAccumulator, consumer, listener, maxNumberOfRetries));
```

Comment:
Can we assume that this consumer is async? Or do we need to put this on another thread? All the uses I know of are async.

Reply:
Yes, it's reasonable to assume that this is lightweight. It's worth documenting that assumption somewhere, so that the caller can take responsibility for forking to another thread if it needs to do something heavier.